// @cloudinary/url-gen/qualifiers/textStyle.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var qualifiers_fontWeight = require('./fontWeight.cjs');
var qualifiers_fontStyle = require('./fontStyle.cjs');
var qualifiers_textDecoration = require('./textDecoration.cjs');
var internal_utils_serializeCloudinaryCharacters = require('../internal/utils/serializeCloudinaryCharacters.cjs');

/**
 * @summary qualifier
 * @description Specifies how to style your layered text, controls the font, font size, line spacing and more.
 * </br><b>Learn more</b>: {@link https://cloudinary.com/documentation/image_transformations#adding_text_overlays|Adding text overlays to images}
 * </br><b>Learn more</b>: {@link https://cloudinary.com/documentation/video_manipulation_and_delivery#adding_text_captions|Adding text overlays to videos}
 * @see {@link Actions.Overlay| The overlay action}
 * @see {@link Actions.Underlay| The underlay action}
 * @memberOf Qualifiers
 */
var TextStyle = /** @class */ (function () {
    /**
     * @param {string} fontFamily The font family
     * @param {number | string} fontSize The font size
     */
    function TextStyle(fontFamily, fontSize) {
        if (!fontFamily || !fontSize) {
            throw "You must provide a fontFamily and fontSize to a TextStyle";
        }
        this._fontFamily = fontFamily;
        this._fontSize = fontSize;
    }
    /**
     * @param {number} spacing The spacing between multiple lines in pixels.
     */
    TextStyle.prototype.lineSpacing = function (spacing) {
        this._lineSpacing = spacing;
        return this;
    };
    /**
     * @param spacing The spacing between the letters, in pixels.
     */
    TextStyle.prototype.letterSpacing = function (spacing) {
        this._letterSpacing = spacing;
        return this;
    };
    /**
     * The antialias setting to apply to the text. When this parameter is not specified, the default antialiasing for the subsystem and target device are applied.
     * @param {FontAntialiasType|string} antiAlias
     */
    TextStyle.prototype.fontAntialias = function (antiAlias) {
        this._fontAntialias = antiAlias;
        return this;
    };
    /**
     * The name of any universally available font or a custom font, specified as the public ID of a raw, authenticated font in your account.
     * For details on custom fonts, see {@link https://cloudinary.com/documentation/image_transformations#using_custom_fonts_for_text_overlays|Using custom fonts for text overlays}.
     * @param {string} fontFamilyName
     */
    TextStyle.prototype.fontFamily = function (fontFamilyName) {
        this._fontFamily = fontFamilyName;
        return this;
    };
    /**
     * @param {number} fontSize The font size
     */
    TextStyle.prototype.fontSize = function (fontSize) {
        this._fontSize = fontSize;
        return this;
    };
    /**
     * @param {FontWeightType|string} fontWeight The font weight
     */
    TextStyle.prototype.fontWeight = function (fontWeight) {
        this._fontWeight = fontWeight;
        return this;
    };
    /**
     *
     * @param {string} fontStyle The font style.
     */
    TextStyle.prototype.fontStyle = function (fontStyle) {
        this._fontStyle = fontStyle;
        return this;
    };
    /**
     * @param {string} fontHinting The outline hinting style to apply to the text. When this parameter is not specified, the default hint style for the font and target device are applied.
     */
    TextStyle.prototype.fontHinting = function (fontHinting) {
        this._fontHinting = fontHinting;
        return this;
    };
    /**
     *
     * @param {TextDecorationType|string} textDecoration The font decoration type.
     */
    TextStyle.prototype.textDecoration = function (textDecoration) {
        this._textDecoration = textDecoration;
        return this;
    };
    /**
     * @param {TextAlignmentType|string} textAlignment The text alignment
     */
    TextStyle.prototype.textAlignment = function (textAlignment) {
        this._textAlignment = textAlignment;
        return this;
    };
    /**
     * @description Whether to include an outline stroke. Set the color and weight of the stroke
     */
    TextStyle.prototype.stroke = function (textStroke) {
        if (textStroke) {
            this._strokeStyle = textStroke;
        }
        this._stroke = true;
        return this;
    };
    TextStyle.prototype.toString = function () {
        return [
            internal_utils_serializeCloudinaryCharacters.serializeCloudinaryCharacters(this._fontFamily) + "_" + this._fontSize,
            this._fontWeight !== qualifiers_fontWeight.normal() && this._fontWeight,
            this._fontStyle !== qualifiers_fontStyle.normal() && this._fontStyle,
            this._textDecoration !== qualifiers_textDecoration.normal() && this._textDecoration,
            this._textAlignment,
            this._stroke && 'stroke',
            this._strokeStyle,
            this._letterSpacing && "letter_spacing_" + this._letterSpacing,
            this._lineSpacing && "line_spacing_" + this._lineSpacing,
            this._fontAntialias && "antialias_" + this._fontAntialias,
            this._fontHinting && "hinting_" + this._fontHinting
        ].filter(function (a) { return a; }).join('_');
    };
    return TextStyle;
}());

exports.TextStyle = TextStyle;
