// @cloudinary/url-gen/actions/layer/LayerAction.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var tslib_es6 = require('../../tslib.es6-18ed4f64.cjs');
var internal_Action = require('../../internal/Action.cjs');
var qualifiers_flag_FlagQualifier = require('../../qualifiers/flag/FlagQualifier.cjs');
var internal_qualifier_Qualifier = require('../../internal/qualifier/Qualifier.cjs');
require('../../internal/utils/dataStructureUtils.cjs');
require('../../internal/models/ActionModel.cjs');
require('../../internal/models/actionToJson.cjs');
require('../../internal/utils/unsupportedError.cjs');
require('../../internal/qualifier/QualifierValue.cjs');
require('../../internal/models/QualifierModel.cjs');
require('../../internal/models/qualifierToJson.cjs');

/**
 * @extends SDK.Action
 * @memberOf SDK
 * @description
 * A generic Layer action that can add a Video, Text or Image layer.<br>
 * This class can represent an overlay or an underlay.
 */
var LayerAction = /** @class */ (function (_super) {
    tslib_es6.__extends(LayerAction, _super);
    /**
     * @description Creates a LayerAction to be used with overlays and underlays
     * @param {ImageSource | TextSource | VideoSource} layerSource The Source used for the layer, use the builders provided {@link Qualifiers.Source|here}
     */
    function LayerAction(layerSource) {
        var _this = _super.call(this) || this;
        _this.source = layerSource;
        return _this;
    }
    /**
     * @description Sets the layerType to 'u' (underlay) or 'l' (overlay).
     * @param {'u' | 'l'} type
     * @return {this}
     */
    LayerAction.prototype.setLayerType = function (type) {
        this.layerType = type;
        return this;
    };
    /**
     * @description Sets the timeline position of the video layer
     * @param {Qualifiers.TimelinePosition} timelinePosition
     * @return {this}
     */
    LayerAction.prototype.timeline = function (timelinePosition) {
        this._timelinePosition = timelinePosition;
        return this;
    };
    /**
     * @description Sets the position of the layer
     * @param {Qualifiers.Position} position
     * @return {this}
     */
    LayerAction.prototype.position = function (position) {
        this._position = position;
        return this;
    };
    /**
     * @description Specifies how to blend the image overlay with the base overlay
     * @param {Qualifiers.BlendMode|BlendModeType} blendMode
     * @return {this}
     */
    LayerAction.prototype.blendMode = function (blendMode) {
        this._blendMode = blendMode;
        return this;
    };
    /**
     * @private
     * @description
     * Closes a layer (layers are built in three stages -> /Open/Transform/Close).
     * @return {SDK.Action}
     */
    LayerAction.prototype.closeLayer = function () {
        var _a, _b, _c, _d;
        var bit = new internal_Action.Action().addFlag(new qualifiers_flag_FlagQualifier.FlagQualifier('layer_apply'));
        (_a = this._position) === null || _a === void 0 ? void 0 : _a.qualifiers.forEach(function (qualifier) {
            bit.addQualifier(qualifier);
        });
        // Flags are stored separately from qualifiers, we need to add those as well
        (_b = this._position) === null || _b === void 0 ? void 0 : _b.flags.forEach(function (flag) {
            bit.addFlag(flag);
        });
        if (typeof this._blendMode === "string") {
            bit.addQualifier(new internal_qualifier_Qualifier.Qualifier('e', this._blendMode));
        }
        else {
            (_c = this._blendMode) === null || _c === void 0 ? void 0 : _c.qualifiers.forEach(function (qualifier) {
                bit.addQualifier(qualifier);
            });
        }
        (_d = this._timelinePosition) === null || _d === void 0 ? void 0 : _d.qualifiers.forEach(function (qualifier) {
            bit.addQualifier(qualifier);
        });
        return bit;
    };
    /**
     * @private
     * @description
     * Opens a layer (layers are built in three stages -> /Open/Transform/Close).
     * @return string
     */
    LayerAction.prototype.openLayer = function () {
        return "" + this.source.getOpenSourceString(this.layerType);
    };
    /**
     * @description
     * Serializes the Layer to a string
     * @return {string}
     */
    LayerAction.prototype.toString = function () {
        return [
            this.openLayer(),
            this.source.getTransformation() && this.source.getTransformation().toString(),
            this.closeLayer()
        ].filter(function (a) { return a; }).join('/');
    };
    return LayerAction;
}(internal_Action.Action));

exports.LayerAction = LayerAction;
